### ISSUE TITLE
Unable to receive packet in virtio_netmap_rxsync()

### ISSUE THREAD

**Issue by anindyasen** [luigirizzo/netmap#327]
When I use the netmap patched virtio_net driver in KVM no packet is received in virtio_netmap_rxsync(). The token from virtqueue_get_buf() is returned NULL.

**Version Information**:
OS : CentOS 7.2
Kernel: Kernel-3.10.0-327.el7.x86_64

**Comment by vmaffione** [comment]
Is your virtio-net device using more than one RX queue? (this is not yet supported)
Is the transmission working at least?
I think virtio patch has never been tried on these old kernels.
But we do use it with newer kernels (4.8+), and it works ok.

**Comment by anindyasen** [comment]
virtio_net device has one RX queue only.
Transmission is also not working may be for the same reason ie. token from virtqueue_get_buf() is returned NULL.
What I observe is even through the virtnet_poll() of virtio_net driver is patched with netmap, packets are not received because the patch returns NM_IRQ_PASS which means driver should do normal interrupt processing as if the interface is not in netmap mode.
So, I checked netmap_do_regif()  that  moves the interface in netmap-mode. Here I find that for the first kring id registration in virtio_netmap_reg is working fine but, for the next kring id it is returning failure as the active fds are more than zero.

**Comment by vmaffione** [comment]
It's not true in general that virtio-net device has one RX queue only. A single TX/RX queue pair is only the default, look at VIRTIO_NET_F_MQ for details.

The virtio-net netmap patch has a limitation, that is it does not support single-queue mode (e.g. /R and /T flags). So this may be your problem. How are you using nm_open() in your program?

**Comment by anindyasen** [comment]
I have 4 threads Rx thread, Tx thread, HostStack Rx thread and HostStack Tx thread. For a interface, there is a nm_open for each thread, ie. 4 nm_open for one interface from 4 different thread.

What do u mean by netmap patch does not support single-queue mode I thought earlier you said netmap is supported for only one Rx queue?Are this single-queue and Rx queue different?

**Comment by vmaffione** [comment]
I was not suggesting to add multiple queues, that wouldn't help.

The failure is not a bug, it is the result of a limitation of the current virtio-net netmap patch: (LINUX/virtio_netmap.h). To avoid more intrusive changes in the virtio-net driver (which is always evolving and hard to maintain), there is a limitation: when you nm_open() a virtio-net device, you must either open all the hardware rings (i.e. all the TX rings and all the RX rings), or open none of the hardware rings. You can instead open host rings separately, that is not a problem.
In other words you cannot open a strict subset of the hardware rings.

Netmap itself supports both opening a single pair of TX/RX queues, or opening all the RX queues (but not the TX queues), or opening all the TX queues (but not the RX queues), etc etc.
The limitation only affects virtio-net devices.

Since you are using netmap with VMs, why don't you use netmap passthrough (ptnet), which is just faster?

**Comment by anindyasen** [comment]
When I nm_open() on a interface, it should opens all the hardware rings (TX Rings and RX Rings), so netmap patch should work for this case.

In single queue mode, the KVM has 1 Rx queue and 1 Tx queue with 256 slots each, which is a single pair of TX/RX queue which netmap supports.
So, virtio-net device should work in my case.

Is there any way check whether all the hardware rings are open by nm_open() or not?

**Comment by vmaffione** [comment]
If you just nm_open("netmap:eth0", ...) it should work. This also means that you should use only one thread.
You can set the verbose module parameter to 1 to see more verbose output, including which rings were opened by netmap.

**Comment by anindyasen** [comment]
I am using CentOS 7.2 OS with Kernel-3.10.0-327.el7.x86_64, and I want use thread 4 thread(Rx, Tx, Hoststack Rx, and Hoststack Tx) to recieve packets through the netmap patched drivers, which is possible with other driver like e1000, e1000e. Is there any way I can implement that in the current netmap source code?

Can you give me some pointer on how I can change the code?

**Comment by vmaffione** [comment]
If you want to run your code in a VM you can use ptnet (see #240 ) instead of virtio.
Then your backend device on the host can be anything (e.g. a physical interface, or a VALE port, or a pipe).

**Comment by anindyasen** [comment]
Actually the problem with ptnet is I dont want to use the modified QEMU, ie. I dont want to change the KVM or the QEMU.

**Comment by vmaffione** [comment]
Only changes to QEMU would be needed.
Anyway, if you want to go through the virtio route, you need to furtherly patch the virtio_net driver in order to
- provide a version of free_unused_bufs() that detaches only the buffers of a queue (RX or TX)
- provide a version of free_receive_bufs() that frees only the buffers of a RX queue
- provide a version of virtio_netmap_clean_used_rings() that gets only the completed buffers of a queue (RX or TX)
- provide a version of virtio_netmap_reclaim_unused() that detaches only the buffers of a queue (RX or TX)
- in the register routine, call the functions above only for the "pending" krings, that is the ones that are switching from normal mode to netmap mode or the other way around (see nm_kring_pending_on() and nm_kring_pending_off()). Right now, the functions above are called for all the TX/RX rings, because we ony accept the case where nm_kring_pending_on() is true for all the hw rings, or nm_kinrg_pending_off() is true for all the hw rings.
